### PR TITLE
A raft of changes to support running GeoMAD

### PR DESCRIPTION
**Pull request by alexgleith**
This PR changes far too many things, so apologies about that.

And I think there are still a range of improvements and tweaks we can make. But, what we have now is working really well, and I think these changes are all a step in the right direction.

Please got through critically, Jessie, and let's maybe workshop it next week to see if we can find a compromise that we can merge in.

### CODE REVIEW

**Comment by jessjaco** [comment]
Just did a quick review. Things look good. I'm thinking we could combine the 2 odc loaders - either add a param `return_type` which defines the return type, or have the existing odcloader subclass the flat loader.  I prefer to not have duplicated code if we can help it.

Happy to help with that part, when the interface is stable.

**Comment by alexgleith** [comment]
Yeah, good idea. Should be easy to do.

Then we can parameterise the structure and returned data type, which should provide some good flexibility.

**Comment by jessjaco** [comment]
I like adding the odc writer. I'm thinking about making a separate mixin for odc vs rioxarray writing (not unlike the structure in the loaders), so as not to have to pass more args through to write_to_blob_storage. Sound good?

**Comment by alexgleith** [comment]
Yeah, that works I guess.

I think there'll be more code duplication that way though, won't there?

**Comment by alexgleith** [comment]
Nice! Although that's still not doing the kind of logging I'm after, I'm sure I can create another variant that does.

We almost need a week long workshop to get this tooling working better!

I'm on hold while I wait for our infrastructure to come back online, but when it has and I have done some test runs of the geomad, I'll come back and break this out into a bunch of smaller PRs... and using your proposed task runner sounds good.

**Comment by jessjaco** [comment]
That's what I was thinking...just make a Task that has whatever logging you want

**Comment by alexgleith** [comment]
> Although that's still not doing the kind of logging I'm after

I have an idea around logging and "have we done this already" checking, which I can implement later.

File: dep_tools/utils.py [hunk 1]
```diff
@@ -43,10 +47,10 @@ def shift_negative_longitudes(
     return LineString([(((pt[0] + 360) % 360), pt[1]) for pt in geometry.coords])
 
 
-@retry(tries=10, delay=1)
-def search_across_180(gpdf: GeoDataFrame, **kwargs) -> ItemCollection:
+@retry(tries=2, delay=1)
+def search_across_180(cell: GeoDataFrame, **kwargs) -> ItemCollection:
     """
-    gpdf: A GeoDataFrame.
+    cell: A GeoDataFrame.
```
Comment (jessjaco) [inline dep_tools/utils.py:53]: `cell` is confusing here. This is a generic function, I  think we should leave the param name generic too

File: dep_tools/utils.py [hunk 2]
```diff
@@ -64,26 +68,32 @@ def search_across_180(gpdf: GeoDataFrame, **kwargs) -> ItemCollection:
...
 
         left_bbox = [xmin_ll, ymin_ll, 180, ymax_ll]
         right_bbox = [-180, ymin_ll, xmax_ll, ymax_ll]
         return ItemCollection(
             list(catalog.search(bbox=left_bbox, **kwargs).items())
             + list(catalog.search(bbox=right_bbox, **kwargs).items())
         )
-
-    return catalog.search(bbox=bbox_4326, **kwargs).item_collection()
+    else:
```
Comment (jessjaco) [inline dep_tools/utils.py:95]: No love for my R-style return logic? I can live with that.

File: dep_tools/utils.py [hunk 3]
```diff
@@ -135,25 +150,39 @@ def write_to_local_storage(
...
+    client: ContainerClient = None,
     **kwargs,
 ) -> None:
-    container_client = get_container_client(**kwargs)
-
-    blob_client = container_client.get_blob_client(str(path))
+    # Allowing for a shared container client, which might be
+    # more efficient. If not provided, get one.
+    if client is None:
+        client = get_container_client()
```
Comment (jessjaco) [inline dep_tools/utils.py:165]: What happened to my kwargs

Comment (jessjaco) [reply dep_tools/utils.py:165]: I guess, if you're going to add client as an arg, just drop the kwargs

File: dep_tools/writers.py [hunk 4]
```diff
@@ -53,31 +56,45 @@ def write(self, xr: Dataset, item_id: str) -> Union[str, List]:
...
-            asset = Asset(
-                media_type="image/tiff; application=geotiff; profile=cloud-optimized",
-                href=str(az_prefix / path),
-                roles=["data"],
-            )
-            assets[variable] = asset
+        client = get_container_client()
+
+        # Use a threadpool to write all at once
+        with ThreadPoolExecutor() as executor:
```
Comment (jessjaco) [inline dep_tools/writers.py:62]: If xr is a lazy dask array (how I try to work in almost all my processing), won't this muck up the scheduler  (since the data aren't processed until write)? 

Also,  writes can occasionally be memory bound and if we're doing them concurrently it could make it worse. 

I also wonder if some of the slowdown you were seeing is because you were testing locally?

Finally, it's also worth considering the processing here, if even generically. If variables within a dataset are derivatives of the same data, then loading variables independently may incur multiple reads of the same source data (because dask / xarray has no way of knowing the data have already been loaded in another variable). Early on when I was performance focused one of the reasons I was writing multiband tiffs was it reduced the number of network reads. I did run into other issues with this approach (I have old notes somewhere), but if you are indeed going to load all the data concurrently maybe just call xr.load() first so it can maybe gain some of these improvements

At the very least, we need a way to opt out of this

File: dep_tools/s2_utils.py [hunk 5]
```diff
@@ -34,26 +36,39 @@ def scale_and_offset_s2(da: DataArray) -> DataArray:
...
+    # VEGETATION = 4
+    # NOT_VEGETATED = 5
+    # WATER = 6
+    # UNCLASSIFIED = 7
+    # CLOUD_MEDIUM_PROBABILITY = 8
     CLOUD_HIGH_PROBABILITY = 9
-    THIN_CIRRUS = 10
-    SNOW = 11
+    # THIN_CIRRUS = 10
+    # SNOW = 11
```
Comment (jessjaco) [inline dep_tools/s2_utils.py:53]: I'm guessing your linter didn't like these unused variables?

File: dep_tools/s2_utils.py [hunk 6]
```diff
@@ -34,26 +36,39 @@ def scale_and_offset_s2(da: DataArray) -> DataArray:
...
+
+    try:
+        cloud_mask = xr.sel(band="SCL").astype("uint16") & bitmask != 0
+    except KeyError:
+        cloud_mask = xr.SCL.astype("uint16") & bitmask != 0
 
     if dilate:
-        clouds = mask_cleanup(clouds, [("opening", 2), ("dilation", 3)])
+        # From Alex @ https://gist.github.com/alexgleith/d9ea655d4e55162e64fe2c9db84284e5
+        cloud_mask = mask_cleanup(cloud_mask, [("opening", 2), ("dilation", 3)])
```
Comment (jessjaco) [inline dep_tools/s2_utils.py:66]: Would be nice, at some point to condense logic that is reused from landsat_utils

**Review by jessjaco — CHANGES_REQUESTED** [review]

File: dep_tools/writers.py (hunk 4 above)
Comment (jessjaco) [reply dep_tools/writers.py:62]: And another thought I had last night: the single band writes of the (multiband) dataset might be why it looks why it's running multiple times

Comment (alexgleith) [reply dep_tools/writers.py:62]: > might be why it looks why it's running multiple times

Good point. But I don't think that is what was happening. The GeoMAD loads everything into memory as part of the processing and also, the dask graph I was watching was loading red, green, blue, ... all the bands.

> At the very least, we need a way to opt out of this

This is probably a good idea. Although, I think it's much more likely to be memory bound earlier in the processing, and from what I have seen, concurrent writing is much more efficient than sequential. Agree it might not be the case for all products. Also I do think forcing an `xr = xr.compute()` is a really good idea just before writing... unless we really optimise dask writing.

File: dep_tools/s2_utils.py (hunk 5 above)
Comment (alexgleith) [reply dep_tools/s2_utils.py:53]: Yeah! We can keep them exposed if you want, it's not a big deal. It _is_ nice documentation having them there!

File: dep_tools/utils.py (hunk 2 above)
Comment (alexgleith) [reply dep_tools/utils.py:95]: Just aiming for readability.

I like having a single return statement normally, so having logic that results in the same return statement running. Having return statements in the middle of a function can be hard to follow.

This change was just me aiming for low cognitive load for future alex (and jesse)!

File: dep_tools/namers.py [hunk 7]
```diff
@@ -26,15 +26,24 @@ def __post_init__(self):
         self._folder_prefix = f"dep_{self.sensor}_{self.dataset_id}/{self.version}"
         self.item_prefix = f"dep_{self.sensor}_{self.dataset_id}"
 
-    def _folder(self, item_id) -> str:
+    def _format_item_id(self, item_id, join_str="/") -> str:
```
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/namers.py:29]: Yeah, I agree.

This was getting a bit messy and opinionated for sure.

I like your solution here.

File: dep_tools/loaders.py [hunk 8]
```diff
@@ -148,35 +174,50 @@ def _get_xr(
         # see in the docs.
         areas_proj = areas.to_crs(self._current_epsg)
         bounds = areas_proj.total_bounds.tolist()
+
+        data_type = "uint16" if self.keep_ints else "float32"
```
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/loaders.py:178]: Since these loaders are specific to Landsat or Sentinel, there's not really going to be any other data types.

If we build a generic product loader, then perhaps we need more flexibility.

File: dep_tools/loaders.py [hunk 9]
```diff
@@ -91,21 +91,38 @@ def _get_items(self, area):
         fix_bad_epsgs(item_collection)
         item_collection = remove_bad_items(item_collection)
 
+        # TODO: these can be queries, which means we don't get them back
+        # from the server in the first place.
```
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/loaders.py:95]: We can do a query like `"query={"platform":{"in":["landsat-9", "landsat-8"]}}"`

Does your syntax work?

The way I've done that query is like `query={"landsat:collection_category": {"in": ["T1"]}}`

File: dep_tools/landsat_utils.py [hunk 10]
```diff
@@ -1,26 +1,39 @@
...
 from pystac import ItemCollection
 from retry import retry
 from xarray import DataArray
 
 
-def mask_clouds(xr: DataArray, dilate: bool = False) -> DataArray:
-    # dilated cloud, cirrus, cloud, cloud shadow
-    mask_bitfields = [1, 2, 3, 4]
+def mask_clouds(
+    xr: DataArray, dilate: Tuple[int, int] | None = None, keep_ints: bool = False
```
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/landsat_utils.py:12]: Yeah, I agree here.

I'm torn between following the way the ODC folks do it (like you suggest) `[["opening", 5], ["dilation", 5]]` or just being opinionated and having True/False.

I think having it changeable is important... so we should use the same API as them. Maybe instead of `None` as a default, we have what we're using, so if you don't want to make a call, it masks clouds the "right" way.

File: dep_tools/namers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/namers.py:29]: The ODC folks have a parameter for region code formatting like `"x{x:03d}y{y:03d}"`

File: dep_tools/loaders.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jessjaco) [reply dep_tools/loaders.py:178]: But this is a generic loader. I've intended it that way and have already used it to load our own data: https://github.com/digitalearthpacific/dep-wofs/blob/4fd9555f37e0e2118ef5230c64662392395c87a6/src/mask.py#L21

File: dep_tools/loaders.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jessjaco) [reply dep_tools/loaders.py:95]: OK, I just wasn't familiar with the syntax enough

File: dep_tools/writers.py (hunk 4 above)
Comment (jessjaco) [reply dep_tools/writers.py:62]: I don't want to enforce a compute, because ideally the tooling should reach a point where it can read, process, and write chunk by chunk. So only a chunk of input and output data is in memory at once. The choking point currently is the need to write to memory before copying to azure, but the new s3 writing for example, offers a way out of that.

I agree about concurrent writing. That's why I was writing multiband tiffs. However, this solution still makes me nervous

File: dep_tools/loaders.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/loaders.py:178]: Oh, you're right sorry.

I guess we should sit and look at this. The `odc-stac` load function has a range of parameters that we should really not be re-implementing.

File: dep_tools/loaders.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/loaders.py:95]: Yeah, it's not easy to discover!

File: dep_tools/loaders.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/loaders.py:178]: The example config from ODC STAC is here at the bottom of the page: https://odc-stac.readthedocs.io/en/latest/_api/odc.stac.load.html

File: dep_tools/landsat_utils.py [hunk 11]
```diff
@@ -1,26 +1,39 @@
...
+    dilation = dilate[1]
+    if dilate is not None:
+        cloud_mask = mask_cleanup(cloud_mask, [("opening", opening), ("dilation", dilation)])
 
-    if dilate:
-        # From Alex @ https://gist.github.com/alexgleith/d9ea655d4e55162e64fe2c9db84284e5
-        cloud_mask = mask_cleanup(cloud_mask, [("opening", 2), ("dilation", 3)])
-    return xr.where(~cloud_mask)
+    if keep_ints:
+        return erase_bad(xr, cloud_mask)
```
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/landsat_utils.py:42]: I think we should ensure we load data with nodata set properly, then we can assume it's set everywhere else.

For this to work we need to be consistent, of course... to be discussed.

File: dep_tools/loaders.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/loaders.py]: Added filter logic and tests.

File: dep_tools/loaders.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/loaders.py:181]: I think we need to leave this one for later.

The best way to handle this is to pass the datatype config straight through to the odc loader and to not do logic around `nan` or `int` masking here, I think.

File: dep_tools/namers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/namers.py:29]: Tweaked this a bit to pass through strings.

We need to refactor to do this right.

File: dep_tools/processors.py [hunk 12]
```diff
@@ -20,14 +21,19 @@ def __init__(
         self,
         send_area_to_processor: bool = False,
         scale_and_offset: bool = True,
-        dilate_mask: bool = False,
+        mask_clouds: bool = True,
+        dilate_mask: Tuple[int, int] | None = None,
+        keep_ints: bool = False
```
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/processors.py:26]: The intention is that this handles ints, but doesn't assume that they are ints.

File: dep_tools/runner.py [hunk 13]
```diff
@@ -37,53 +37,41 @@ def __init__(self, tasks: GeoDataFrame, continue_on_error: bool = True, **kwargs
         super().__init__(tasks, **kwargs)
         self.continue_on_error = continue_on_error
 
+    def run_one(self, index):
```
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/runner.py:40]: Ok, will remove these changes.

File: dep_tools/utils.py (hunk 1 above)
Comment (alexgleith) [reply dep_tools/utils.py:52]: Changed "cell" to "region"

File: dep_tools/utils.py (hunk 3 above)
Comment (alexgleith) [reply dep_tools/utils.py:164]: Good point. Added kwargs back in for the container client getting.

File: dep_tools/writers.py (hunk 4 above)
Comment (alexgleith) [reply dep_tools/writers.py:62]: Ok, added a fallback.

The multi-threaded writing worked brilliantly though! Simple and effective :-)

File: dep_tools/landsat_utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (alexgleith) [reply dep_tools/landsat_utils.py:42]: I'll lift this out as an issue.

File: dep_tools/task.py [hunk 14]
```diff
@@ -46,7 +48,8 @@ def run(self):
             dict(area=self.area) if self.processor.send_area_to_processor else dict()
         )
         output_data = self.processor.process(input_data, **processor_kwargs)
-        self.writer.write(output_data, self.id)
+        paths = self.writer.write(output_data, self.id, multithreaded=self.multithreaded)
```
Comment (jessjaco) [inline dep_tools/task.py:51]: We can't do this, multithreaded is not part of the Writer class

File: dep_tools/writers.py [hunk 15]
```diff
@@ -49,35 +52,66 @@ class DsWriter(XrWriterMixin, Writer):
     write_stac_function: Callable = write_stac_blob_storage
     write_stac: bool = True
 
-    def write(self, xr: Dataset, item_id: str) -> Union[str, List]:
+    def write(self, xr: Dataset, item_id: str, multithreaded: bool = False) -> Union[str, List]:
```
Comment (jessjaco) [inline dep_tools/writers.py:55]: `multithreaded` should be a class var or it breaks the abstraction

File: dep_tools/writers.py [hunk 16]
```diff
@@ -49,35 +52,66 @@ class DsWriter(XrWriterMixin, Writer):
...
+            for variable in xr:
+                output_da = xr[variable].squeeze()
+                path = self.itempath.path(item_id, variable)
+                paths.append(path)
+
+                self.write_function(
+                    output_da,
+                    path=path,
+                    write_args=dict(
+                        driver="COG", nodata=output_da.attrs.get("nodata", None)
```
Comment (jessjaco) [inline dep_tools/writers.py:94]: This is already the default value for nodata, see https://odc-geo.readthedocs.io/en/latest/_api/odc.geo.xr.to_cog.html

Comment (jessjaco) [reply dep_tools/writers.py:94]: Also, I will remove the driver="COG" default on next go around, I want to sound a warning

**Comment by jessjaco** [comment]
Ok, I made some changes which needed to happen and some to improve clarity. I'm still not totally settled on the loader in particular, but also would like to refactor the write_to_blob_storage function. However, I think these can be best handled piecewise

**Comment by alexgleith** [comment]
Looks clean! Nice work, especially with the partial function. That's neat.

**Comment by jessjaco** [comment]
Pending any further input, I'd like to merge this tomorrow and relegate fixes or additional changes to new PRs

**Comment by alexgleith** [comment]
Happy for you to merge it in.

File: dep_tools/namers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jessjaco) [reply dep_tools/namers.py:29]: I need to fix this. Can you tell me what things you want to pass in so I can test them all? I'm guessing 
"1,2" [should be "001_002"]
("1,2", FJ) [should be "001_002_FJ"] <- are you still doing this?

I need to pass in e.g. (1,2) and it's not working
